### PR TITLE
Add app metadata to all events

### PR DESCRIPTION
**Pull request by Guardiola31337**
### Description

Adds app metadata to all events

### Changelog

```
<changelog>Added optional `EventsAppMetadata` field to `NavigationOptions`.</changelog>
<changelog>Exposed `NavigationSessionState` and `NavigationSessionStateObserver`.</changelog>
<changelog>Removed `AppMetadata` from `MapboxNavigation.postUserFeedback`.</changelog>
```

Opening it as a `Draft` to discuss if we're good with the approach taken.

### CODE REVIEW

**Comment by Guardiola31337** [comment]
cc'ing @bamx23 @jill-cardamon for visibility on the iOS side

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt [hunk 1]
```diff
@@ -71,10 +69,24 @@ internal class NavigationSession : RoutesObserver, TripSessionStateObserver {
...
+
+    /**
+     * Free Drive state
+     */
+    FREE_DRIVE,
+
+    /**
+     * Active Guidance state
+     */
+    ACTIVE_GUIDANCE
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:91]: We should be more specific here and describe the "requirements" for entering or leaving any of these states. A diagram would be great as well.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt [hunk 2]
```diff
@@ -180,16 +185,19 @@ internal object MapboxNavigationTelemetry {
         when (sessionState) {
             IDLE, FREE_DRIVE -> {
                 sessionStop()
+                eventsSessionId = obtainUniversalUniqueIdentifier()
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:188]: This can be moved up from `when` as it's the same for all cases. Unless we'd like to clear it for `IDLE`?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt [hunk 3]
```diff
@@ -71,10 +69,24 @@ internal class NavigationSession : RoutesObserver, TripSessionStateObserver {
...
 
-    internal enum class State {
-        IDLE,
-        FREE_DRIVE,
-        ACTIVE_GUIDANCE
-    }
+/**
+ * Contains the various states that can occur during a navigation.
+ */
+enum class NavigationSessionState {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:77]: Should this be a sealed class and have the session ID as a value?

File: examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt [hunk 4]
```diff
@@ -198,9 +197,8 @@ class MapboxNavigationActivity : AppCompatActivity() {
         }
     }
 
-    private val routeProgressObserver = object : RouteProgressObserver {
-        override fun onRouteProgressChanged(routeProgress: RouteProgress) {
-            // update the camera position to account for the progressed fragment of the route
+    private val routeProgressObserver =
+        RouteProgressObserver { routeProgress -> // update the camera position to account for the progressed fragment of the route
```
Comment (kmadsen) [inline examples/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt:201]: nit:
```suggestion
        RouteProgressObserver { routeProgress -> 
        // update the camera position to account for the progressed fragment of the route
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt (hunk 3 above)
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:77]: Came here to say this too. if the session id is in the observable state, we could delete `EventsSessionIdObserver` and only observe the `NavigationSession`.

We could surface a `NavigationSession` which has a `State` and an `id`. The class responsible for figuring this out would be `NavigationSessionController` (right now it's the `NavigationSession`)

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt [hunk 5]
```diff
@@ -180,16 +185,19 @@ internal object MapboxNavigationTelemetry {
...
                 locationsCollector.flushBuffers()
-                handleStateChanged(this.sessionState, sessionState)
                 needStartSession = true
+                handleStateChanged(this.sessionState, sessionState)
+                eventsSessionId = obtainUniversalUniqueIdentifier()
                 startSessionIfNeedAndCan()
             }
         }
         this.sessionState = sessionState
+        eventsSessionIdObserver?.onEventsSessionIdChanged(eventsSessionId)
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:200]: Is it possible for the state to change multiple times for the same navigation session?

For example, does the sessionId change for every route change? Does this mean re-routes create new sessions? It looks like this observer is triggered whenever the state is _updated_, not just when it _changes_

https://github.com/mapbox/mapbox-navigation-android/blob/ca57002edb30b82ed2cb9c0eb3180d6beda951fa/libnavigation-core/src/main/java/com/mapbox/navigation/core/NavigationSession.kt#L20

**Comment by Guardiola31337** [comment]
Addressed the feedback @kmadsen @LukasPaczos this is ready for another round of 👀

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    /**
+     * Random session UUID.
+     * This is generated internally based on the current state within an trip session.
+     * I.e. will change when transitioning across states of a trip session. Empty when [Idle].
+     *
+     * Useful to use it in combination with the [MapboxHistoryRecorder].
+     *
+     * @see [TripSessionState]
+     */
+    var sessionId: String = ""
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:119]: why do we need it here?

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:119]:
```kotlin
* Useful to use it in combination with the [MapboxHistoryRecorder].
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt (hunk 2 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt]: What do you mean? BTW that's outdated. `sessionId` is returned as part of the `NavigationSessionState` from `NavigationSessionStateObserver`.

Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt]: get it 👍

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/options/EventsAppMetadata.kt [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        return true
+    }
+
+    /**
+     * Regenerate whenever a change is made
+     */
+    override fun hashCode(): Int {
+        var result = name.hashCode()
+        result = 31 * result + version.hashCode()
+        result = 31 * result + (userId?.hashCode() ?: 0)
```
Comment (RingerJK) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/options/EventsAppMetadata.kt:52]:
```suggestion
        result = 31 * result + userId.hashCode()
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+ * - [FreeDrive]
+ * - [ActiveGuidance]
+ *
+ * The SDK starts off in an [Idle] state.
+ * Whenever the [MapboxNavigation.startTripSession] is called, the SDK will enter the [FreeDrive] state.
+ * If the session is stopped, the SDK will enter the [Idle] state.
+ * If the SDK is in an [Idle] state, it stays in this same state even when a primary route is available.
+ * If the SDK is already in the [FreeDrive] mode or entering it whenever a primary route is available,
+ * the SDK will enter the [ActiveGuidance] mode instead.
+ * When the routes are manually cleared, the SDK automatically fall back to either [Idle] or [FreeDrive] state.
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:106]:
```suggestion
 * When the routes are manually cleared, the SDK automatically fall back to [FreeDrive] state.
 * If a state is already [ActiveGuidance] and a new route is set, the state and [sessionId] are not changed.
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 9]
```diff
@@ -879,25 +883,31 @@ class MapboxNavigation(
...
     }
 
     /**
-     * Register a [NavigationSessionStateObserver] to be notified of the various Session states. Not publicly available
+     * Register a [NavigationSessionStateObserver] to be notified of the various Session states.
+     *
+     * @param navigationSessionStateObserver NavigationSessionStateObserver
      */
-    internal fun registerNavigationSessionObserver(
+    fun registerNavigationSessionObserver(
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:899]:
```suggestion
    fun registerNavigationSessionStateObserver(
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 10]
```diff
@@ -879,25 +883,31 @@ class MapboxNavigation(
...
     }
 
     /**
-     * Unregisters a [NavigationSessionStateObserver]. Not publicly available
+     * Unregisters a [NavigationSessionStateObserver].
+     *
+     * @param navigationSessionStateObserver NavigationSessionStateObserver
      */
-    internal fun unregisterNavigationSessionObserver(
+    fun unregisterNavigationSessionObserver(
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:910]:
```suggestion
    fun unregisterNavigationSessionStateObserver(
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt [hunk 11]
```diff
@@ -178,14 +181,14 @@ internal object MapboxNavigationTelemetry {
...
+            Idle, FreeDrive -> {
                 sessionStop()
                 handleStateChanged(this.sessionState, sessionState)
             }
-            ACTIVE_GUIDANCE -> {
+            ActiveGuidance -> {
                 locationsCollector.flushBuffers()
-                handleStateChanged(this.sessionState, sessionState)
                 needStartSession = true
+                handleStateChanged(this.sessionState, sessionState)
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:191]: Why the change of order?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt [hunk 12]
```diff
@@ -125,15 +127,16 @@ internal object MapboxNavigationTelemetry {
     private val feedbackEventCacheMap = LinkedHashMap<String, NavigationFeedbackEvent>()
 
     private var needHandleReroute = false
-    private var sessionState: NavigationSession.State = IDLE
+    private var sessionState: NavigationSessionState = Idle
     private var routeProgress: RouteProgress? = null
     private var originalRoute: DirectionsRoute? = null
     private var needStartSession = false
+    private var eventsSessionId: String = ""
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:134]: This is never set as far as I can tell. Shouldn't we assign it when state changes?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+     *
+     * @see [TripSessionState]
+     */
+    var sessionId: String = ""
+        internal set
+
+    /**
+     * Idle state
+     */
+    object Idle : NavigationSessionState() {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:126]: Making this can object will overwrite IDs of past sessions, won't it?

If we get a state with an ID, and then the state goes through a cycle, the ID of our old cached value will be different.

This should rather be a class
```suggestion
    class Idle(sessionId: String) : NavigationSessionState(sessionId) {
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 9 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:899]: Since the session state now also carries dynamic session ID information, we should expose a `NavigationSessionState` getter as well.

**Comment by Guardiola31337** [comment]
@LukasPaczos ready for a final review 🚀

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 14]
```diff
@@ -393,6 +398,15 @@ class MapboxNavigation(
...
     }
 
+    /**
+     * Provides the current navigation session state.
+     *
+     * @see NavigationSessionState
+     * @see NavigationSessionStateObserver
+     * @see [registerNavigationSessionStateObserver]
+     */
+    val navigationSessionState = navigationSession.state
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:408]: This is a variable because of the underlying implementation of `navigationSession.state`, not an immutable value as the declaration would suggest. I'd suggest making this a function:
```suggestion
    fun getNavigationSessionState(): NavigationSessionState = navigationSession.state
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt [hunk 15]
```diff
@@ -178,13 +181,15 @@ internal object MapboxNavigationTelemetry {
     private val navigationSessionStateObserver = NavigationSessionStateObserver { sessionState ->
         log("session state is $sessionState")
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:182]:
```suggestion
        log("session state is $sessionState")
        this.sessionState = sessionState
```
Otherwise, some of the methods below would use the old, outdated field value. Would be great to cover this with a unit test as well.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt:182]: Clarified that this setup is correct for the old schema, due to be updated in https://github.com/mapbox/mapbox-navigation-android/pull/4693.

**Review by LukasPaczos — APPROVED** [review]
:ship:

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * Random session UUID.
+     * This is generated internally based on the current state within an trip session.
+     * I.e. will change when transitioning across states of a trip session. Empty when [Idle].
+     *
+     * Useful to use it in combination with the [MapboxHistoryRecorder].
+     *
+     * @see [TripSessionState]
+     */
+    abstract var sessionId: String
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:114]: should be?
```suggestion
    abstract val sessionId: String
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        internal set
+
+    /**
+     * Idle state
+     */
+    data class Idle internal constructor(
+        override var sessionId: String
+    ) : NavigationSessionState()
```
Comment (RingerJK) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:122]:
```suggestion
    object Idle internal : NavigationSessionState() {
       override val sessionId = ""
    }
```

**Review by RingerJK — CHANGES_REQUESTED** [review]
a few questions

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt (hunk 16 above)
Comment (RingerJK) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/NavigationSession.kt:114]: I don't see that it might be changed, do it?

**Review by RingerJK — APPROVED** [review]
